narayana/ArjunaJTA/jta/tests/classes/io/narayana/perf/product/ProductComparison.java

```diff
@@ -130,8 +129,7 @@ protected void doWork(ProductInterface product) throws Exception {
         } catch (Exception e) {
             if (errorCount.incrementAndGet() > MAX_ERRORS) {
                 e.printStackTrace();
-
-                throw new BenchmarkException(e);
+                Thread.currentThread().interrupt();
```

Comment:
Please can you say why the interrupt is better than the exception?

Reply:
Here in the this link http://mail.openjdk.java.net/pipermail/jmh-dev/2016-September/002333.html . They mentioned that, there was an issues (with three different scenarios) with benchmark, which they have fixed. 

The issue that we are facing looks similar to first mentioned scenario i.e. 
>a) If the benchmark is throwing the exception, we still unconditionally sleep for the iteration time; we should instead ask all threads to stop, and throw the exception early;

Currently we are using JMH version 1.21 which was released in May, 2018. So I think it should have the fix(fix was given in Sept 2016).
Seems like the issue is not resolved yet.

So I thought to try other way to terminate the benchmark.

Reply:
I wonder if there is a newer version of JMH we can use? What is the current version?

Reply:
I think interrupt as discussed in that link related to what JMH is trying to do to the benchmark, rather than a way to get the benchmark to terminate

Reply:
> I wonder if there is a newer version of JMH we can use? What is the current version?

JMH version 1.23 is the latest- released in Jan 2020

Reply:
> I think interrupt as discussed in that link related to what JMH is trying to do to the benchmark, rather than a way to get the benchmark to terminate

I didn't understand

Reply:
Ah! Okay, I got it what you mean. But I was wondering, if we could interrupt it, instead of them interrupting it.

Reply:
I think the interrupt issued by ourselves would not be so useful as we are the thing running on the thread at this time? A thread interrupting itself doesn't seem to help?

I do think we should try the upgrade to the latest JMH version

Reply:
Okay, sure we can do that, if it resolves the issue.

Reply:
I think it would be useful to consider the upgrade anyway

Reply:
Before these changes, I tried throwing ExecutionException (because, here benchmark compution is involved), in build http://narayanaci1.eng.hst.ams2.redhat.com/job/btny-pulls-performance/10/consoleFull , we see that, though it failed, but it got out of the for loop exception. 

This exception looping is only happening when Benchmarked with 1600 thread count.

Now the 11th run is with thread.interrupt() changes.

Reply:
Now I am thinking, its better to throw ExecutionException instead of BenchmarkException

Reply:
Very interesting re the different exceptions! I wonder what the API for JMH has to say about the difference. Re: interrupt I still don't really see how interrupting ourselves being the right route forwards but I could be wrong

Reply:
I am not very clear on why ExecutionException was selected?

Reply:
BenchmarkException seems to be an internal exception, might be relevant:
https://github.com/openjdk/jmh/blob/master/jmh-core/src/main/java/org/openjdk/jmh/runner/BenchmarkException.java#L31 (at least in the openjdk GitHub read-only mirror)

Reply:
okay